core/sr-primitives/src/generic/tip.rs

```diff
@@ -0,0 +1,47 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Tip structure for a transaction.
+
+use crate::codec::{Encode, Decode};
+
+/// Representation of a transaction tip.
+///
+/// Upon decoding, all transaction types try and decode this from the end of the encoded byte
+/// stream.
+/// If non-existent, the default implementation will be used.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Copy, Eq, PartialEq, Encode, Decode)]
+pub enum Tip<Balance> {
```

Comment:
Does the tip balance really need to be generic? Can't we just require it's something that converts to `Balance` like `u64` (I believe via `SimpleArithmetic` we should be able to cast to `Balance` anyway)

Reply:
Yeah, I actually had in mind to make it bound to `SimpleArithmetic` at some point. For now and for the sake of speed in proceeding, I just assumed it is _anything_.

Looks good? this is what I had in mind: 

```suggestion
pub enum Tip<N: SimpleArithmetic> {
```

Reply:
Ok, I misunderstood what you meant. I can make it a `type TipCurrency = u64`. This can be converted into a balance using `SaturatingFrom` but using that is highly discouraged __unless if saturating makes sense__. I don't see why it should make sense to cap all tips to u64 and hence I would still argue that it should be a generic, preferably bounded to `SimpleArithmetic` for clarity.

Reply:
> unless if saturating makes sense

I don't think the `Tip` should ever saturate `Balance` type - that would be nuts, first acquiring all available currency and then spending it for a tip.

`Tip` will always be strictly smaller than any possible `Balance` of any account, so if it ever saturates and assuming we check that you have enough funds to spend the tip it just means the transaction is invalid or the chain is broken (since someone has saturated the amount of `Balance` he has).
It may include some maleability (i.e. if `Balance < u64` transaction with `Tip = Balance::max_value()` and `Tip = Balance::max_value() + 1` would have different has, but essentially would be processed in the same way, but again it should always be just non-spendable)

I'm totally fine with having it generic, but for code simplification we could consider bounding it to some more reasonable value. The only issue I see is if `Balance` is a super huge type and we expect `Tip`s to be higher than `u64`.